File: experiment/cf1609b/main.cpp
// AC

#include <bits/stdc++.h>
using namespace std;

#define ALL(X) (X).begin(), (X).end()
#define FOR(I, X, Y) for (auto I = decltype(Y){X}; I < (Y); ++I)
using ll = int64_t;
using ull = uint64_t;
namespace std {
template<class ...Ts>        istream& operator>>(istream& i,       tuple<Ts...>& x) { apply([&](auto&&... ys){ ((i >> ys), ...); }, x); return i; }
template<class ...Ts>        ostream& operator<<(ostream& o, const tuple<Ts...>& x) { o << "("; auto s = ""; apply([&](auto&&... y){ ((o << s << y, s = ", "), ...); }, x); return o << ")"; }
template<class T1, class T2> istream& operator>>(istream& i,       pair<T1, T2>& x) { return i >> tie(x.first, x.second); }
template<class T1, class T2> ostream& operator<<(ostream& o, const pair<T1, T2>& x) { return o << tie(x.first, x.second); }
template<class T, class = decltype(begin(declval<T>())), class = enable_if_t<!is_same<T, string>::value>>
istream& operator>>(istream& i,       T& x) { for (auto& y : x) { i >> y; } return i; }
template<class T, class = decltype(begin(declval<T>())), class = enable_if_t<!is_same<T, string>::value>>
ostream& operator<<(ostream& o, const T& x) { o << "{"; auto s = ""; for (auto& y : x) { o << s << y; s = ", "; } return o << "}"; }
}

#ifdef DEBUG
#define dbg(...) do { cerr << #__VA_ARGS__ ": " << make_tuple(__VA_ARGS__) << endl; } while (0)
#define dbg2(X) do { cerr << #X ":\n"; for (auto& __x : (X)) { cout << __x << endl; } } while (0)
#else
#define dbg(...)
#define dbg2(X)
#endif

const string kPattern = "abc";

// Main
void mainCase() {
  int n, nq; // [1, 10^5]
  cin >> n >> nq;
  string s;
  cin >> s;
  vector<tuple<int, char>> qs(nq);
  cin >> qs;

  // Count matching segment
  int count = 0;

  // Initial state
  for (int i = 0; i < n; i++) {
    if (s.substr(i, 3) == kPattern) {
      count++;
    }
  }

  // Incrementally update
  for (auto [i, c] : qs) {
    i--;

    // Check only three segments where possibly "abc" appeared or disappeared
    for (int j = max(i - 2, 0); j <= i; j++) {
      count -= (s.substr(j, 3) == kPattern);
    }

    s[i] = c;

    for (int j = max(i - 2, 0); j <= i; j++) {
      count += (s.substr(j, 3) == kPattern);
    }

    cout << count << "\n";
  }
}

int main() {
  ios_base::sync_with_stdio(0); cin.tie(0);
  mainCase();
  return 0;
}

/*
python misc/run.py experiment/cf1609b/main.cpp

%%%% begin
9 10
abcabcabc
1 a
1 b
2 c
3 a
4 b
5 c
8 a
9 b
1 c
4 a
%%%%
3
2
2
2
1
2
1
1
1
0
%%%% end
*/

File: experiment/cf1609b/main.py
# AC

import os
from sys import stderr, stdin, stdout

DEBUG = os.getenv("DEBUG")
PATTERN = bytearray("abc", "ascii")


def main_case() -> None:
    n, q = map(int, stdin.readline().split())  # [1, 10^5]
    s = bytearray(input(), "ascii")

    # Count initial state
    count = 0
    for i in range(n - 2):
        if s[i : i + 3] == PATTERN:
            count += 1

    # Process queries
    for _ in range(q):
        i_, c = stdin.readline().split()
        i = int(i_) - 1

        # Check three ranges where possibly "abc" appeared or disappeared
        #   ? ? X ? ?
        #   <--->
        #     <--->
        #       <--->
        for j in range(i - 2, i + 1):
            if s[j : j + 3] == PATTERN:
                count -= 1

        s[i] = ord(c)
        if DEBUG:
            print(s, file=stderr)
            print(count, file=stderr)

        for j in range(i - 2, i + 1):
            if s[j : j + 3] == PATTERN:
                count += 1

        stdout.write(f"{count}\n")


def main() -> None:
    main_case()


if __name__ == "__main__":
    main()


"""
python -m experiment.cf1609b.main < experiment/cf1609b/in.txt

%%%% begin
9 10
abcabcabc
1 a
1 b
2 c
3 a
4 b
5 c
8 a
9 b
1 c
4 a
%%%%
3
2
2
2
1
2
1
1
1
0
%%%% end
"""
